### PR TITLE
Remove `with_options!` macro

### PR DESCRIPTION
**Pull request by sploiselle**
Implemented enough option handling to remove the `with_options!` macro.

### Motivation

This PR adds a known-desirable feature. Part of #5390

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - there are no user-facing behavior changes

### CODE REVIEW

File: src/sql-parser/src/parser.rs [hunk 1]
```diff
@@ -3238,6 +3250,22 @@ impl<'a> Parser<'a> {
...
 
+    fn parse_copy_options(&mut self) -> Result<CopyOption<Raw>, ParserError> {
+        let name =
+            match self.expect_one_of_keywords(&[FORMAT, DELIMITER, NULL, ESCAPE, QUOTE, HEADER])? {
+                FORMAT => CopyOptionName::Format,
+                DELIMITER => CopyOptionName::Format,
+                NULL => CopyOptionName::Format,
+                ESCAPE => CopyOptionName::Format,
+                QUOTE => CopyOptionName::Format,
+                HEADER => CopyOptionName::Format,
```
Comment (benesch) [inline src/sql-parser/src/parser.rs:3261]:
```suggestion
                FORMAT => CopyOptionName::Format,
                DELIMITER => CopyOptionName::Delimiter,
                NULL => CopyOptionName::Null,
                ESCAPE => CopyOptionName::Escape,
                QUOTE => CopyOptionName::Quote,
                HEADER => CopyOptionName::Header,
```

File: src/sql/src/plan/with_options.rs [hunk 2]
```diff
@@ -140,10 +149,13 @@ impl<V: TryFromValue<Value>> ImpliedValue for Vec<V> {
...
 }
 
-impl<V: TryFromValue<Value>, T: AstInfo> TryFromValue<WithOptionValue<T>> for V {
+impl<V: TryFromValue<Value> + StrConv, T: AstInfo + std::fmt::Debug>
+    TryFromValue<WithOptionValue<T>> for V
+{
     fn try_from_value(v: WithOptionValue<T>) -> Result<Self, anyhow::Error> {
         match v {
             WithOptionValue::Value(v) => V::try_from_value(v),
+            WithOptionValue::Ident(i) => Ok(V::strconv(i.as_str())?),
```
Comment (benesch) [inline src/sql/src/plan/with_options.rs:158]: Still digging in, but it seems like we can avoid this and the `StrConv` trait and associated macros with something like:

```
WithOptionValue::Ident(i) => V::try_from_value(Value::string(i.as_str())),
```

**Review by benesch — COMMENTED** [review]
Getting so close! 🙌🏽

File: src/sql/src/plan/with_options.rs (hunk 2 above)
Comment (sploiselle) [reply src/sql/src/plan/with_options.rs:158]: Definitely--when I woke up this morning, I had the same idea.

File: src/sql-parser/src/parser.rs (hunk 1 above)
Comment (sploiselle) [reply src/sql-parser/src/parser.rs:3261]: Derp!

**Review by benesch — APPROVED** [review]
This is just so good.

**Pull request merged**